cmd/thanos/sidecar.go

```diff
@@ -49,6 +50,30 @@ func registerSidecar(m map[string]setupFunc, app *kingpin.Application, name stri
 	gcsBucket := cmd.Flag("gcs.bucket", "Google Cloud Storage bucket name for stored blocks. If empty sidecar won't store any block inside Google Cloud Storage").
 		PlaceHolder("<bucket>").String()
 
+	s3Bucket := cmd.Flag("s3.bucket", "S3-Compatible API bucket name for stored blocks.").
+		PlaceHolder("<bucket>").String()
+
+	s3Endpoint := cmd.Flag("s3.endpoint", "S3-Compatible API endpoint for stored blocks.").
+		PlaceHolder("<api-url>").String()
+
+	s3AccessKey := cmd.Flag("s3.access-key", "Access key for an S3-Compatible API.").
+		PlaceHolder("<key>").String()
+
+	s3SecretKey := cmd.Flag("s3.secret-key", "Secret key for an S3-Compatible API.").
+		PlaceHolder("<key>").String()
```

Comment:
Is this _the_ or _one_ authentication option for S3? In Prometheus upstream, the service discovery has at least 3 ways of doing it and IIRC, at least one request to add yet another one. I find this extremely confusing and compicated – how are the odds that we get away with something simpler in Thanos :)?

Reply:
I 100% agree that this is confusing and overcomplicated, but I'm not aware of any better way.

With the qualification that I don't know a _ton_ about S3 and AWS in general, it seems there are other ways to authenticate, using stored credentials (`(~/.aws/credentials)`) or using EC2 Instance Role Credentials so you can just auto-auth if you're running Thanos on EC2. Or I think there's other role-like features.

The simplest path I think would be to allow these things to be supplied as environment variables. It seems like a pretty pervasive pattern that if you use S3 you've always got your secret/access keys hanging around. It looks like [kingpin can do that](https://godoc.org/github.com/alecthomas/kingpin#Application.DefaultEnvars).

Otherwise, we could certainly collapse these four flags into one somehow. I don't know that it's common but I doubt it's unprecedented. `--s3.bucket accesskey:secretkey@url/bucket` or something. But if people are used to dealing with them individually that's an extra step for building this custom string in automation.

Another thing to consider when using a very specific Amazon authentication tool (IAMs/Roles and the like) is that it might make it more difficult for folks to use this storage backend for services other than S3. As difficult as it is, the s3 API is implemented by a lot of different services, so this driver using the standard, albeit janky interface allows it to remain generic enough to work with a multitude of services (Minio and Spaces, [among others](http://www.s3-client.com/s3-compatible-storage-solutions.html)).

Reply:
IMO environment variables will be ok here. This would be (kind of) consistent with out current approach for GCS secret. We don't specify any special flag for it, we just leave it to the GCS client which expects JSON file in `GOOGLE_APPLICATION_CREDENTIALS`. 

The only downside here is that it could be hard for the users to deduct how to specify these secrets. Maybe worth to just document them better, in comment for example?

Reply:
Yea, there's certainly a lot of value in things to just work in a VM that has the appropriate env vars already pre-set. But S3 being basically the generic object storage API, we probably need to support multiple options.

Maybe we shouldn't worry for now and iterate based on user feedback now. Nothing is declared stable yet anyway.